sql: v21.2.4: unsupported OID 2249 with format code FormatText [using record type in extended protocol Bind]

This issue was autofiled by Sentry. It represents a crash or reported error on a live cluster with telemetry enabled.

Sentry link: https://sentry.io/organizations/cockroach-labs/issues/2999516817/?referrer=webhooks_plugin

Panic message:

>encoding.go:799: error in argument for ×: unsupported OID 2249 with format code FormatText
--
*errutil.leafError: unsupported OID 2249 with format code FormatText (1)
encoding.go:799: *withstack.withStack (top exception)
*assert.withAssertionFailure
*errutil.withPrefix: error in argument for × (2)
conn_executor_prepare.go:397: *withstack.withStack (3)
*pgerror.withCandidateCode: 08P01 (4)
(check the extra data payloads)


<details>
<summary>Stacktrace (expand for inline code snippets):</summary>

https://github.com/cockroachdb/cockroach/blob/0c8df44947208da2f6234d2736e8ab6d3b0937c0/pkg/sql/conn_executor_prepare.go#L396-L398 in pkg/sql.(*connExecutor).execBind
https://github.com/cockroachdb/cockroach/blob/0c8df44947208da2f6234d2736e8ab6d3b0937c0/pkg/sql/conn_executor.go#L1805-L1807 in pkg/sql.(*connExecutor).execCmd
https://github.com/cockroachdb/cockroach/blob/0c8df44947208da2f6234d2736e8ab6d3b0937c0/pkg/sql/conn_executor.go#L1630-L1632 in pkg/sql.(*connExecutor).run
https://github.com/cockroachdb/cockroach/blob/0c8df44947208da2f6234d2736e8ab6d3b0937c0/pkg/sql/conn_executor.go#L669-L671 in pkg/sql.(*Server).ServeConn
https://github.com/cockroachdb/cockroach/blob/0c8df44947208da2f6234d2736e8ab6d3b0937c0/pkg/sql/pgwire/conn.go#L647-L649 in pkg/sql/pgwire.(*conn).processCommandsAsync.func1
/usr/local/go/src/runtime/asm_amd64.s#L1370-L1372 in runtime.goexit
</details>

```
pkg/sql/conn_executor_prepare.go in pkg/sql.(*connExecutor).execBind at line 397
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).execCmd at line 1806
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).run at line 1631
pkg/sql/conn_executor.go in pkg/sql.(*Server).ServeConn at line 670
pkg/sql/pgwire/conn.go in pkg/sql/pgwire.(*conn).processCommandsAsync.func1 at line 648
/usr/local/go/src/runtime/asm_amd64.s in runtime.goexit at line 1371
```

| Tag | Value |
|---|---|
| Cockroach Release | `v21.2.4` |
| Cockroach SHA: | 0c8df44947208da2f6234d2736e8ab6d3b0937c0 |
| Platform | linux amd64 |
| Distribution | CCL |
| Environment | development |
| Command | server |
| Go Version | ``|
| # of CPUs ||
| # of Goroutines ||


Jira issue: CRDB-13129

--- a/pkg/sql/pgwire/pgwirebase/encoding.go
+++ b/pkg/sql/pgwire/pgwirebase/encoding.go
@@ -317,6 +317,12 @@ func DecodeDatum(
 	switch code {
 	case FormatText:
 		switch id {
+		case oid.T_record:
+			d, _, err := tree.ParseDTupleFromString(evalCtx, string(b), t)
+			if err != nil {
+				return nil, err
+			}
+			return d, nil
 		case oid.T_bool:
 			t, err := strconv.ParseBool(string(b))
 			if err != nil {

--- a/pkg/sql/pgwire/testdata/pgtest/tuple
+++ b/pkg/sql/pgwire/testdata/pgtest/tuple
@@ -64,3 +64,30 @@ ReadyForQuery
 {"Type":"DataRow","Values":[{"text":"(a,b,c,\"d \",e,\"f  \")"}]}
 {"Type":"CommandComplete","CommandTag":"SELECT 1"}
 {"Type":"ReadyForQuery","TxStatus":"I"}
+
+# Try to send a prepared statement with a tuple argument.
+# 'S' for Statement
+# ParameterFormatCodes = [0] for text format
+send
+Parse {"Name": "s4", "Query": "select $1 AS a", "ParameterOIDs": [2249]}
+Bind {"DestinationPortal": "p4", "PreparedStatement": "s4", "ParameterFormatCodes": [0], "Parameters": [{"text":"(1,cat)"}]}
+Execute {"Portal": "p4"}
+Sync
+----
+
+# Postgres has a slightly different error message.
+until noncrdb_only keepErrMessage
+ErrorResponse
+ReadyForQuery
+----
+{"Type":"ParseComplete"}
+{"Type":"ErrorResponse","Code":"0A000","Message":"input of anonymous composite types is not implemented"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+until crdb_only keepErrMessage
+ErrorResponse
+ReadyForQuery
+----
+{"Type":"ParseComplete"}
+{"Type":"ErrorResponse","Code":"0A000","Message":"error in argument for $1: could not parse \"(1,cat)\" as type tuple: cannot parse anonymous record type"}
+{"Type":"ReadyForQuery","TxStatus":"I"}

